velox/substrait/SubstraitExtensionCollector.h

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <optional>
+#include "velox/core/Expressions.h"
+#include "velox/core/PlanNode.h"
+#include "velox/substrait/VeloxSubstraitSignature.h"
+#include "velox/substrait/proto/substrait/algebra.pb.h"
+#include "velox/substrait/proto/substrait/plan.pb.h"
+#include "velox/type/Type.h"
+
+namespace facebook::velox::substrait {
+
+struct ExtensionFunctionId {
+  /// Substrait extension YAML file uri.
+  std::string uri;
+
+  /// Substrait signature used in the function extension declaration is a
+  /// combination of the name of the function along with a list of input
+  /// argument types.The format is as follows : <function
+  /// name>:<short_arg_type0>_<short_arg_type1>_..._<short_arg_typeN> for more
+  /// detail information about the argument type please refer to link
+  /// https://substrait.io/extensions/#function-signature-compound-names.
+  std::string signature;
+
+  bool operator==(const ExtensionFunctionId& other) const {
+    return (uri == other.uri && signature == other.signature);
+  }
+};
+
+/// Assigns unique IDs to function signatures using ExtensionFunctionId.
+class SubstraitExtensionCollector {
+ public:
+  SubstraitExtensionCollector();
+
+  /// Given a scalar function name and argument types, return the functionId
+  /// using ExtensionFunctionId.
+  int getReferenceNumber(
+      const std::string& functionName,
+      const std::vector<TypePtr>& arguments);
+
+  /// Given an aggregate function name and argument types and aggregation Step,
+  /// return the functionId using ExtensionFunctionId.
+  int getReferenceNumber(
+      const std::string& functionName,
+      const std::vector<TypePtr>& arguments,
+      core::AggregationNode::Step aggregationStep);
+
+  /// Add extension functions to Substrait plan.
+  void addExtensionsToPlan(::substrait::Plan* plan) const;
+
+ private:
+  /// A bi-direction hash map to keep the relation between reference number and
+  /// either function or type signature.
+  /// @T ExtensionFunctionId
+  template <class T>
+  class BiDirectionHashMap {
+   public:
+    /// If the forwardMap_ does not contain the key, then the key and value will
```

Comment:
This behavior seems difficult to work with. For the bi-map, users would expect that the map enforces uniqueness of both keys and values, no?

Consider, what happens when we add the following:

1. 1, 10 => {1 -> 10}, {10 ->1}
1. 1, 20 => {1 -> 20}, {10 ->1, 20 -> 1}
1. 2, 20 => {1 -> 20, 2 -> 20}, {10 -> 1, 20 -> 2}